model.py

```diff
@@ -0,0 +1,118 @@
+import pandas as pd
+import matplotlib.pyplot as plt
+import numpy as np
+
+from sklearn.model_selection import train_test_split
+from sklearn.ensemble import RandomForestRegressor
+from sklearn.metrics import mean_squared_error
+
+from preprocessing import preprocessing
+
+all_data = preprocessing()
+
+# print(all_data)
+'''
+    Location        DATE    Alamosa     Denver      Grand    Springs
+    0         2003-01-01  24.985507  33.523256  26.692130  31.296296
+    1         2003-01-02  15.527778  37.068966  27.435374  31.213178
+    2         2003-01-03  22.891156  47.344828  28.850340  43.343537
+    3         2003-01-04  26.864583  39.812500  31.074830  38.122449
+    4         2003-01-05  30.284722  41.914286  32.323529  39.490000
+    ...              ...        ...        ...        ...        ...
+    7300      2022-12-27  31.413793  48.848485  31.000000  49.303571
+    7301      2022-12-28  36.068966  39.130435  32.466292  41.100000
+    7302      2022-12-29  21.800000  25.578947  24.492754  28.648649
+    7303      2022-12-30  20.327586  26.424242  28.638889  27.965517
+    7304      2022-12-31  30.800000  32.794118  31.682540  44.338710
+    '''
+
+all_data['DATE'] = pd.to_datetime(all_data['DATE'])
+
+#Adding lag and rolling mean features (Feature engineering to help with the models accuracy)
+lag_featues = pd.DataFrame()
+rolling_mean_feature = pd.DataFrame()
+rolling_median_feature = pd.DataFrame()
+rolling_std_feature = pd.DataFrame()
+differential_feature = pd.DataFrame()
+
+# Shift() function in pandas creates these shifted or lag features from time series data.
+# Adding a lag of 1 day
+for lag in range(1,2):
+    for location in ['Alamosa','Grand','Springs','Denver']:
+        lag_featues[f"{location}_lag_{lag}"] = all_data[location].shift(lag)
+
+# Trying a few different rolling window sizes
+window_sizes = [1,3,5,7]
+
+for window_size in window_sizes:
+    for location in ['Alamosa','Grand','Springs','Denver']:
+        rolling_mean_feature[f"{location}_rolling_mean_{window_size}"] = all_data[location].rolling(window=window_size).mean()
+        rolling_std_feature[f"{location}_rolling_std_{window_size}"] = all_data[location].rolling(window=window_size).std()
+        rolling_median_feature[f"{location}_rolling_median_{window_size}"] = all_data[location].rolling(window=window_size).median()
+
+# Differential features
+for location in ['Alamosa','Grand','Springs','Denver']:
+    differential_feature[f"{location}_diff"] = all_data[location].diff()
+
+# Seasonality features
+all_data['day_of_year'] = all_data['DATE'].dt.dayofyear
+all_data['month'] = all_data['DATE'].dt.month

```

Comment:
I just commented these line out and ran the model to see if would significantly affect its performance and these were the results:

RMSE: 0.13149971446300704
                     Feature    Importance
8       Grand_rolling_mean_1  5.313442e-01
24    Grand_rolling_median_1  4.686337e-01
40       Grand_rolling_std_3  5.121341e-06
44       Grand_rolling_std_5  2.370562e-06
30   Denver_rolling_median_3  1.153345e-06
38   Denver_rolling_median_7  7.694791e-07
28    Grand_rolling_median_3  6.836496e-07
43     Alamosa_rolling_std_5  6.046888e-07
6               Denver_lag_1  6.041270e-07
3              Alamosa_lag_1  5.858034e-07
32    Grand_rolling_median_5  5.676044e-07
52                Grand_diff  5.059094e-07
15    Alamosa_rolling_mean_5  4.804759e-07
54               Denver_diff  4.066776e-07
14     Denver_rolling_mean_3  3.759358e-07
13    Springs_rolling_mean_3  3.425429e-07
4                Grand_lag_1  3.344873e-07
22     Denver_rolling_mean_7  3.286346e-07
46      Denver_rolling_std_5  3.233396e-07
5              Springs_lag_1  3.209218e-07
12      Grand_rolling_mean_3  3.002498e-07
47     Alamosa_rolling_std_7  2.991868e-07
31  Alamosa_rolling_median_5  2.802270e-07
25  Springs_rolling_median_1  2.790084e-07
34   Denver_rolling_median_5  2.773095e-07
10     Denver_rolling_mean_1  2.768552e-07
21    Springs_rolling_mean_7  2.728729e-07
7     Alamosa_rolling_mean_1  2.474804e-07
33  Springs_rolling_median_5  2.273338e-07
29  Springs_rolling_median_3  2.233874e-07
50      Denver_rolling_std_7  2.146506e-07
18     Denver_rolling_mean_5  2.069187e-07
53              Springs_diff  2.059068e-07
36    Grand_rolling_median_7  2.038909e-07
35  Alamosa_rolling_median_7  1.886106e-07
51              Alamosa_diff  1.850115e-07
49     Springs_rolling_std_7  1.733119e-07
1                     Denver  1.562074e-07
20      Grand_rolling_mean_7  1.546598e-07
0                    Alamosa  1.505968e-07
48       Grand_rolling_std_7  1.445823e-07
41     Springs_rolling_std_3  1.431983e-07
45     Springs_rolling_std_5  1.431536e-07
37  Springs_rolling_median_7  1.239353e-07
39     Alamosa_rolling_std_3  1.152274e-07
42      Denver_rolling_std_3  1.116691e-07
17    Springs_rolling_mean_5  1.102809e-07
26   Denver_rolling_median_1  1.059195e-07
9     Springs_rolling_mean_1  1.029977e-07
23  Alamosa_rolling_median_1  1.012801e-07
19    Alamosa_rolling_mean_7  1.006775e-07
2                    Springs  1.000123e-07
27  Alamosa_rolling_median_3  9.834569e-08
16      Grand_rolling_mean_5  9.667500e-08
11    Alamosa_rolling_mean_3  9.332801e-08


It looks like they did not significantly affect the model performance. I think the model is able to pick up on this seasonal patterns from the lag and rolling window features

Reply:
Nice. Given that the model seems to find the pattern of seasons via some other model feature(s), maybe it may be worth separating this out in the future. You must be right that the rolling window/lag features (or other features) must be catching the knowledge that the seasons (independently) bring.